target/linux/ramips/dts/rt5350_dlink_dir-506l.dts

```diff
@@ -0,0 +1,159 @@
+/dts-v1/;
+
+#include "rt5350.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "dlink,dir-506l", "ralink,rt5350-soc";
+	model = "D-Link DIR-506L";
+
+	aliases {
+		led-boot = &led_status_green;
+		led-failsafe = &led_status_red;
+		led-running = &led_status_green;
+		led-upgrade = &led_status_red;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		usb {
+			label = "dir-506l:green:usb";
+			gpios = <&gpio0 9 GPIO_ACTIVE_LOW>;
+		};
+
+		led_status_green: status_green {
+			label = "dir-506l:green:status";
+			gpios = <&gpio0 19 GPIO_ACTIVE_HIGH>;
+		};
+
+		led_status_red: status_red {
+			label = "dir-506l:red:status";
+			gpios = <&gpio1 1 GPIO_ACTIVE_HIGH>;
+		};
+
+		internet {
+			label = "dir-506l:green:internet";
+			gpios = <&gpio1 0 GPIO_ACTIVE_HIGH>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
```

Comment:
Can you try with gpio-keys for compatible and dropping poll-interval?

Reply:
I check other rt5350 based devices dts files, they all use ```compatible = "gpio-keys-polled";```.

Reply:
Newly added keenetic lite b uses it. https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/rt5350_zyxel_keenetic-lite-b.dts#L35
And I'm pretty sure it'll also work on your device.

Reply:
Ok, I will make build with this and spi change and test on real h/w.
But I remember that polling required for this chip, do not remember where I read this.

Reply:
You are using battery status gpios as keys here, and I'm not sure it will fit in this case... Typically keys are for state transition (except for EV_SW).

Reply:
https://oldwiki.archive.openwrt.org/doc/howto/hardware.button
actions: pressed, timeout, released

Reply:
Keys are for buttons, and battery status gpios are not.
Does the kernel get signal if the gpio is high before it boots?
IMO this should be dealt with gpio-export (though I'm not sure it will work with led-gpio). You don't use rc.button script for this anyway, do you?

Reply:
I'm not linux kernel developer and have no idea what happen during boot.

Does project have any rule to deny use GPIO input from non physical buttons as buttons?

Personally I do not use any additional scripts in owrt, but other users may simple hook this "buttons" to send alert on power fail/restore, low power, charge complete, etc...

Reply:
One of my concern is that it's difficult to know its status before the status changes. (Keys are not exported to sysfs) Personally I would use a custom script reading gpio-export regularly and turning on/off the leds accordingly.

Reply:
https://openwrt.org/docs/guide-user/hardware/hardware.button - Sliding switches
https://github.com/jefferyto/openwrt-slide-switch
this can be handled by scripts.
TP-Link TL-MR3020 and some others use GPIO input as switch.

Reply:
Yeah, TL-MR3020 has a real hardware slide switch...

Reply:
To me, using gpio-export here also seems more desirable if it can be done properly.

Reply:
Is any disadvantage for this, except that this is not physical button/switch?
Is there any devices that uses GPIO input from non button/switch in base to use it as example?
For me is clear how to use this buttons, and examples exist in wifi, and usage of generic GPIO input is not.

Reply:
The disadvantage is that it's a hack. These are no buttons, so they shouldn't be implemented as such.
I won't merge it this way.